urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/FileExtractor.java

```diff
@@ -0,0 +1,218 @@
+package com.kokakiwi.mclauncher.core.updater;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Enumeration;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import net.contrapunctus.lzma.LzmaInputStream;
+
+import com.kokakiwi.mclauncher.api.LauncherAPI;
+import com.kokakiwi.mclauncher.core.updater.UpdaterWorker.GameFile;
+
+public class FileExtractor
+{
+    public static boolean extract(LauncherAPI api, GameFile source, File dest,
+            int min, int max, boolean recursive) throws Exception
+    {
+        final String[] exts = source.getFileName()
+                .substring(source.getFileName().indexOf('.', 0) + 1)
+                .split("\\.");
+        
+        File file = source.getFile();
+        
+        for (int i = exts.length - 1; i > -1; i--)
+        {
+            final String ext = exts[i];
+            boolean pass = false;
+            
+            if (ext.equalsIgnoreCase("lzma"))
+            {
+                file = extractLZMA(api, source, file, dest, min, max);
+            }
+            
+            if (ext.equalsIgnoreCase("zip"))
+            {
+                extractZIP(api, source, file, dest, min, max);
+                pass = true;
+            }
+            
+            if (ext.equalsIgnoreCase("jar"))
+            {
+                extractJAR(api, source, file, dest, min, max);
+                pass = true;
+            }
+            
+            if (!recursive || pass)
+            {
+                i = -1;
+            }
+        }
+        
+        return false;
+    }
+    
+    public static File extractLZMA(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        final File uncompressed = new File(dest, file.getName().substring(0,
+                file.getName().lastIndexOf('.')));
+        
+        final FileInputStream fileInput = new FileInputStream(file);
+        
+        InputStream in = new LzmaInputStream(fileInput);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *in*. The resource is referenced at lines 70, 76, 80. The resource is closed at line 83. There are other execution paths that do not contain closure statements, for example, when FileOutputStream constructor throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).